search_is_some fixit hint suggests invalid code

`clippy 0.0.212 (9897442 2019-04-23)`

in this code
````rust
    // https://docs.rs/walkdir/2.2.7/walkdir/#example-skip-hidden-files-and-directories-on-unix
    fn is_git(entry: &walkdir::DirEntry) -> bool {
        entry
            .path()
            .components()
            .find(|&path_elm| path_elm == std::path::Component::Normal(OsStr::new(".git")))
            .is_some()
    }
````
clippy suggests:
````
warning: called `is_some()` after searching an `Iterator` with find. This is more succinctly expressed by calling `any()`.
   --> src/main.rs:352:9
    |
352 | /         entry
353 | |             .path()
354 | |             .components()
355 | |             .find(|&path_elm| path_elm == std::path::Component::Normal(OsStr::new(".git")))
356 | |             .is_some()
    | |______________________^
    |
    = note: #[warn(clippy::search_is_some)] on by default
    = note: replace `find(|&path_elm| path_elm == std::path::Component::Normal(OsStr::new(".git"))).is_some()` with `any(|&path_elm| path_elm == std::path::Component::Normal(OsStr::new(".git")))`
    = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#search_is_some
````

if I apply the suggestion
````rust
    fn is_git(entry: &walkdir::DirEntry) -> bool {
        entry
            .path()
            .components()
            .any(|&path_elm| path_elm == std::path::Component::Normal(OsStr::new(".git")))
    }
````
the code no longer compiles
````
    Checking raca v0.1.0 (/home/matthias/vcs/github/raca)
error[E0308]: mismatched types
   --> src/main.rs:355:19
    |
355 |             .any(|&path_elm| path_elm == std::path::Component::Normal(OsStr::new(".git")))
    |                   ^^^^^^^^^ expected enum `std::path::Component`, found reference
    |
    = note: expected type `std::path::Component<'_>`
               found type `&_`
    = help: did you mean `path_elm: &std::path::Component<'_>`?

error: aborting due to previous error
````

The correct code is
````rust
.any(|path_elm| ...
````
instead of
````rust
.any(|&path_elm| ...
````

--- a/clippy_lints/src/methods/mod.rs
+++ b/clippy_lints/src/methods/mod.rs
@@ -1975,6 +1975,13 @@ fn lint_search_is_some<'a, 'tcx>(
         );
         let search_snippet = snippet(cx, search_args[1].span, "..");
         if search_snippet.lines().count() <= 1 {
+            // suggest `any(|x| ..)` instead of `any(|&x| ..)` for `find(|&x| ..).is_some()`
+            let any_search_snippet =
+                if search_method == "find" && search_snippet.starts_with("|&") {
+                    Some(search_snippet.replacen('&', "", 1))
+                } else {
+                    None
+                };
             // add note if not multi-line
             span_note_and_lint(
                 cx,
@@ -1983,8 +1990,9 @@ fn lint_search_is_some<'a, 'tcx>(
                 &msg,
                 expr.span,
                 &format!(
-                    "replace `{0}({1}).is_some()` with `any({1})`",
-                    search_method, search_snippet
+                    "replace `{0}({1}).is_some()` with `any({2})`",
+                    search_method, search_snippet,
+                    any_search_snippet.as_ref().map_or(&*search_snippet, String::as_str)
                 ),
             );
         } else {

--- a/clippy_lints/src/methods/mod.rs
+++ b/clippy_lints/src/methods/mod.rs
@@ -1977,10 +1977,16 @@ fn lint_search_is_some<'a, 'tcx>(
         if search_snippet.lines().count() <= 1 {
             // suggest `any(|x| ..)` instead of `any(|&x| ..)` for `find(|&x| ..).is_some()`
             let any_search_snippet =
-                if search_method == "find" && search_snippet.starts_with("|&") {
-                    Some(search_snippet.replacen('&', "", 1))
-                } else {
-                    None
+                if_chain! {
+                    if search_method == "find";
+                    if let hir::ExprKind::Closure(_, _, body_id, ..) = search_args[1].node;
+                    let closure_body = cx.tcx.hir().body(body_id);
+                    if let hir::PatKind::Ref(..) = closure_body.arguments[0].pat.node;
+                    then {
+                        Some(search_snippet.replacen('&', "", 1))
+                    } else {
+                        None
+                    }
                 };
             // add note if not multi-line
             span_note_and_lint(

--- a/clippy_lints/src/methods/mod.rs
+++ b/clippy_lints/src/methods/mod.rs
@@ -1980,7 +1980,8 @@ fn lint_search_is_some<'a, 'tcx>(
                 if search_method == "find";
                 if let hir::ExprKind::Closure(_, _, body_id, ..) = search_args[1].node;
                 let closure_body = cx.tcx.hir().body(body_id);
-                if let hir::PatKind::Ref(..) = closure_body.arguments[0].pat.node;
+                if let Some(closure_arg) = closure_body.arguments.get(0);
+                if let hir::PatKind::Ref(..) = closure_arg.pat.node;
                 then {
                     Some(search_snippet.replacen('&', "", 1))
                 } else {

--- a/clippy_lints/src/methods/mod.rs
+++ b/clippy_lints/src/methods/mod.rs
@@ -1976,18 +1976,17 @@ fn lint_search_is_some<'a, 'tcx>(
         let search_snippet = snippet(cx, search_args[1].span, "..");
         if search_snippet.lines().count() <= 1 {
             // suggest `any(|x| ..)` instead of `any(|&x| ..)` for `find(|&x| ..).is_some()`
-            let any_search_snippet =
-                if_chain! {
-                    if search_method == "find";
-                    if let hir::ExprKind::Closure(_, _, body_id, ..) = search_args[1].node;
-                    let closure_body = cx.tcx.hir().body(body_id);
-                    if let hir::PatKind::Ref(..) = closure_body.arguments[0].pat.node;
-                    then {
-                        Some(search_snippet.replacen('&', "", 1))
-                    } else {
-                        None
-                    }
-                };
+            let any_search_snippet = if_chain! {
+                if search_method == "find";
+                if let hir::ExprKind::Closure(_, _, body_id, ..) = search_args[1].node;
+                let closure_body = cx.tcx.hir().body(body_id);
+                if let hir::PatKind::Ref(..) = closure_body.arguments[0].pat.node;
+                then {
+                    Some(search_snippet.replacen('&', "", 1))
+                } else {
+                    None
+                }
+            };
             // add note if not multi-line
             span_note_and_lint(
                 cx,
@@ -1997,7 +1996,8 @@ fn lint_search_is_some<'a, 'tcx>(
                 expr.span,
                 &format!(
                     "replace `{0}({1}).is_some()` with `any({2})`",
-                    search_method, search_snippet,
+                    search_method,
+                    search_snippet,
                     any_search_snippet.as_ref().map_or(&*search_snippet, String::as_str)
                 ),
             );

--- a/tests/ui/methods.stderr
+++ b/tests/ui/methods.stderr
@@ -155,7 +155,7 @@ LL |     let _ = v.iter().find(|&x| *x < 0).is_some();
    |             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
    |
    = note: `-D clippy::search-is-some` implied by `-D warnings`
-   = note: replace `find(|&x| *x < 0).is_some()` with `any(|&x| *x < 0)`
+   = note: replace `find(|&x| *x < 0).is_some()` with `any(|x| *x < 0)`
 
 error: called `is_some()` after searching an `Iterator` with find. This is more succinctly expressed by calling `any()`.
   --> $DIR/methods.rs:236:13

